src/bookshelf/src/bksw_sql.erl

```diff
@@ -0,0 +1,274 @@
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+list_buckets() ->
+    case sqerl:select(list_buckets, [], rows_as_records, ?DB_BUCKET_TX_FM) of
+        {ok, Buckets} ->
+            Buckets;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+list_bucket(BucketName) ->
+    case sqerl:select(list_bucket_with_hash, [BucketName], rows_as_records, ?DB_FILE_TX_FM) of
+        {ok, Buckets} ->
+            Buckets;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+-spec create_file(binary(), binary()) -> integer().
+create_file(Bucket, Name) ->
+    case sqerl:select(insert_file, [Bucket, Name], first_as_scalar, [data_id]) of
+        {ok, File} ->
+            {ok, File};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+
+create_file_with_data(BucketId, FileName, DataId) ->
+    case sqerl:statement(insert_file_with_data_id, [BucketId, FileName, DataId], count) of
+        {ok, 1} ->
+            ok;
+        Error ->
+            Error
+    end.
+
+update_file_with_data(FileId, DataId) ->
+    case sqerl:statement(update_file_with_data_id, [FileId, DataId], count) of
+        {ok, 1} ->
+            ok;
+        Error ->
+            Error
+    end.
+
+find_file(Bucket, Name) ->
+    case sqerl:select(find_file, [Bucket, Name], {first_as_record, ?DB_FILE_TX_FM }) of
+        %% Awkward sanity check that we got back the expected record type here.
+        {ok, Object} when db_file =:= element(1, Object) ->
+            {ok, Object};
+        {ok, none} ->
+            {ok, not_found};
+        {error, Error} ->
+            {error, Error}
+    end.
+
+find_file(FileId) ->
+    case sqerl:select(find_file_by_id, [FileId], {first_as_record, ?DB_FILE_TX_FM }) of
+        %% Awkward sanity check that we got back the expected record type here.
+        {ok, Object} when db_file =:= element(1, Object) ->
+            {ok, Object};
+        {ok, none} ->
+            {ok, not_found};
+        {error, Error} ->
+            {error, Error}
+    end.
+
+%% TODO IMPROVE
+delete_file(Bucket, Name) ->
+    case find_file(Bucket, Name) of
+        {ok, none} ->
+            {error, file_not_found};
+        {ok, #db_file{file_id = FileId}} ->
+            delete_file(FileId);
+        Error ->
+            Error
+    end.
+
+
+delete_file(FileId) ->
+    case sqerl:statement(delete_file, [FileId], count) of
+        {ok, 1} ->
+            ok;
+        Error ->
+            Error
+    end.
+
+rename_file_with_overwrite(BucketId, FileId, NewName) ->
+    case sqerl:statement(rename_file_with_overwrite, [BucketId, FileId, NewName], sucess) of
+        {ok, true} ->
+            ok;
+        Error ->
+            Error
+    end.
+
+mark_file_done(DataId, Size, Chunks, SumMD5, SumSha256, SumSha512) ->
+    case sqerl:statement(update_file_data_done, [DataId, Size, Chunks, SumMD5, SumSha256, SumSha512], count) of
+        {ok, 1} ->
+            ok;
+        Error ->
+            Error
```

Comment:
what happens to the file data if `update_file_data_done` fails? Should there be a way to manage orphaned file data?